### ISSUE TITLE
Datadog Agent Source Regression in v0.24x

### ISSUE THREAD

**Issue by jonwinton** [vectordotdev/vector#15292]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Problem

We noticed the following issue when upgrading from `v0.23.3` to `v0.24.0`:

**(1) CPU became spikey/less consistent**

The charts below show how CPU explodes, Datadog forwarder error rates climb, HAProxy 5xx rates climb

![Screenshot 2022-11-17 at 3 24 52 PM](https://user-images.githubusercontent.com/4906712/202581210-f74f6ad8-ab91-4239-a530-9ddc2785186b.png)

**(2) 504 error codes coming from the Datadog Agents writing to Vector for only the `/api/beta/sketches` endpoint**
```
2022-11-15 22:52:12 UTC | CORE | ERROR | (pkg/forwarder/worker.go:184 in process) | Error while processing transaction: error "504 Gateway Time-out" while sending transaction to "http://vector-haproxy.vector.svc.cluster.local:6000/api/beta/sketches", rescheduling it: "<html><body><h1>504 Gateway Time-out</h1>\nThe server didn't respond in time.\n</body></html>\n"
```
This is also apparent in errors surfacing from HAProxy (deployed via the Vector Helm chart). HAProxy is using a `leastconn` balance strategy.

**(3) Error logs coming from Vector for shutting down connections**
```
{"host":"vector-599576bd9b-w2bq7","message":"error shutting down IO: Transport endpoint is not connected (os error 107)","metadata":{"kind":"event","level":"DEBUG","module_path":"hyper::proto::h1::conn","target":"hyper::proto::h1::conn"},"pid":1,"source_ty
pe":"internal_logs","timestamp":"2022-11-16T20:11:46.533762489Z"}
{"host":"vector-599576bd9b-w2bq7","message":"connection error: error shutting down connection: Transport endpoint is not connected (os error 107)","metadata":{"kind":"event","level":"DEBUG","module_path":"hyper::server::server::new_svc","target":"hyper::se
rver::server::new_svc"},"pid":1,"source_type":"internal_logs","timestamp":"2022-11-16T20:11:46.533777847Z"}
```

### Configuration

```text
data_dir: /vector-data-dir

api:
  enabled: true
  address: 127.0.0.1:8686
  playground: false

sources:
  internal_logs:
    type: internal_logs

  # Datadog Agent telemetry
  datadog_agent:
    type: datadog_agent
    address: "0.0.0.0:6000"
    multiple_outputs: true # To automatically separate metrics and logs

sinks:
  console:
    type: console
    inputs:
      - internal_logs
    target: stdout
    encoding:
      codec: json

  # Datadog metrics output
  datadog_metrics:
    type: datadog_metrics
    inputs:
      - <inputs>...
    api_key: "${DATADOG_API_KEY}"
```

### Version

0.24.0-distroless-libc

### Debug Output

```text
I can only recreate this issue in critical environments where I can't create this output information :(
```

### Example Data

I'm not sure what the Datadog Agent is sending to this endpoint

### Additional Context

We're running in AWS EKS 1.21 in

**Comment by barieom** [comment]
We're looking into this, will keep you posted, @jonwinton !

**Comment by jonwinton** [comment]
Thanks @barieom! I wanted to add this screenshot of steady state CPU usage. Orange line is CPU requests, red line is CPU limit. Looking at network traffic there is are no outliers in terms of higher network load on a single pod.

<img width="1468" alt="Screenshot 2022-11-19 at 8 37 09 AM" src="https://user-images.githubusercontent.com/4906712/202861658-283fa842-54a5-4d99-a276-98c41a95d18e.png">

Let me know if I can supply anything else to help!

**Comment by neuronull** [comment]
Thanks for filing this issue @jonwinton.

It looks like the vector helm chart was used for configuration of vector.

Would you be able to provide the complete helm chart values used, and the chart version you are using?

This would expedite the investigation and reproduction effort.

**Comment by jonwinton** [comment]
Definitely @neuronull !

**Chart Version:** 0.14.0
One comment on this: we're blocked from upgrading due to our current EKS version not having the autoscaling APIs defined in [current versions of the chart](https://github.com/vectordotdev/helm-charts/blob/develop/charts/vector/templates/_helpers.tpl#L81-L90). We're in the progress of updating, but it'll be awhile

We use Argo for deployment of the chart but we have a fun Helm args combining pipeline, I need some time to get the full config together.

**Comment by jonwinton** [comment]
@neuronull here we go: 

```
affinity:
  podAntiAffinity:
    preferredDuringSchedulingIgnoredDuringExecution:
      - podAffinityTerm:
          labelSelector:
            matchExpressions:
            - key: app.kubernetes.io/name
              operator: In
              values:
              - vector
          topologyKey: failure-domain.beta.kubernetes.io/zone
        weight: 100
autoscaling:
  enabled: true
  targetCPUUtilizationPercentage: 80
  targetMemoryUtilizationPercentage: 60
commonLabels:
  tags.datadoghq.com/service: vector
  tags.datadoghq.com/version: "0.24.0"
env:
- name: DATADOG_API_KEY
  valueFrom:
    secretKeyRef:
      key: ...
      name: ...
- name: DD_HOSTNAME
  valueFrom:
    fieldRef:
      fieldPath: spec.nodeName
- name: VECTOR_POD_NAME
  valueFrom:
    fieldRef:
      fieldPath: metadata.name
- name: VECTOR_VERSION
  valueFrom:
    fieldRef:
      fieldPath: metadata.labels['tags.datadoghq.com/version']
extraVolumeMounts:
- mountPath: /mnt/secrets-store
  name: vector-secrets
  readOnly: true
extraVolumes:
- csi:
    driver: secrets-store.csi.k8s.io
    readOnly: true
    volumeAttributes:
      secretProviderClass: ...
  name: vector-secrets
podDisruptionBudget:
  enabled: true
  minAvailable: 10%
podLabels:
  tags.datadoghq.com/service: vector
podPriorityClassName: system-cluster-critical
role: Stateless-Aggregator
rollWorkload: true
serviceAccount:
  annotations:
    eks.amazonaws.com/role-arn: <role id>
  create: true
updateStrategy:
  rollingUpdate:
    maxSurge: 0
    maxUnavailable: 1
  type: RollingUpdate
haproxy:
  enabled: true
  image:
    tag: &version 2.6.6
  resources:
    requests:
      cpu: 1000m
      memory: 1Gi
    limits:
      cpu: 1000m
      memory: 1Gi

  autoscaling:
    enabled: true
    minReplicas: 9
    maxReplicas: 50

  podLabels:
    tags.datadoghq.com/service: vector-haproxy
    tags.datadoghq.com/version: *version

  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
        - podAffinityTerm:
            labelSelector:
              matchExpressions:
              - key: app.kubernetes.io/name
                operator: In
                values:
                - vector
            topologyKey: failure-domain.beta.kubernetes.io/zone
          weight: 100

  podAnnotations:
    ad.datadoghq.com/haproxy.checks: |
      {
        "haproxy": {
          "init_config": {},
          "instances": [{
            "use_openmetrics": true,
            "openmetrics_endpoint": "http://%%host%%:1024/metrics",
            "histogram_buckets_as_distributions": true,
            "collect_counters_with_distributions": true
          }]
        }
      }

  customConfig: |
    global
      log stdout format raw local0
      maxconn 4096
      stats socket /tmp/haproxy
      hard-stop-after {{ .Values.haproxy.terminationGracePeriodSeconds }}s

    defaults
      log     global
      option  dontlognull
      retries 10
      option  redispatch
      option  allbackups
      timeout client 10s
      timeout server 10s
      timeout connect 5s

    resolvers coredns
      nameserver dns1 kube-dns.kube-system.svc.cluster.local:53
      resolve_retries 3
      timeout resolve 2s
      timeout retry 1s
      accepted_payload_size 8192
      hold valid 10s
      hold obsolete 60s

    frontend stats
      mode http
      bind :::1024
      option httplog
      http-request use-service prometheus-exporter if { path /metrics }
      stats enable
      stats hide-version  # Hide HAProxy version
      stats realm Haproxy\ Statistics  # Title text for popup window
      stats uri /haproxy_stats  # Stats URI
      stats refresh 10s


    frontend datadog-agent
      mode http
      bind :::6000
      option httplog
      option dontlog-normal
      default_backend datadog-agent

    backend datadog-agent
      mode http
      balance leastconn
      option tcp-check
      server-template srv 10 _datadog-agent._tcp.{{ include "vector.fullname" $ }}.{{ $.Release.Namespace }}.svc.cluster.local resolvers coredns check
```

**Comment by spencergilbert** [comment]
> **Chart Version:** 0.14.0 One comment on this: we're blocked from upgrading due to our current EKS version not having the autoscaling APIs defined in [current versions of the chart](https://github.com/vectordotdev/helm-charts/blob/develop/charts/vector/templates/_helpers.tpl#L81-L90). We're in the progress of updating, but it'll be awhile

Hey @jonwinton IIRC there's nothing _really_ stopping us from backporting the HPA options to older API's. I think I stopped where I did because there was some amount of structural change that happened between API versions and I figured if someone needed the support we'd get an issue opened.

We'd be happy to get an issue opened for that support if that would help you upgrade.

**Comment by neuronull** [comment]
> @neuronull here we go:

Thanks @jonwinton  !

Forgot to ask as well- what version of the Datadog Agent are you using? Was it also upgraded or did it's version stay the same? Can you share the DD Agent config as well?

Cheers~

**Comment by jonwinton** [comment]
Sorry for the delay @neuronull and @spencergilbert ! Had some holiday time off and everything. Let me get you that info.
---
@neuronull I did test setting `store_api_key` and it didn't help.

**Comment by jonwinton** [comment]
> Forgot to ask as well- what version of the Datadog Agent are you using? Was it also upgraded or did it's version stay the same? Can you share the DD Agent config as well?

@neuronull the version of the DataDog agent has been locked to `7.39.1` for the duration of this test. 

> We'd be happy to get an issue opened for that support if that would help you upgrade.

@spencergilbert I think we're going to be stuck on the autoscaling/v1 API for the next 3-6 months, so if it's possible to support those versions that would be amazing 🙇

**Comment by jszwedko** [comment]
@jonwinton I see you mentioned you tried setting `store_api_key`. To clarify, did you try setting it to `false`? That is: `store_api_key: false` on the `datadog_agent` source. As @neuronull pointed out there was a fix in 0.24.0 to start using the API key that Agent sent metrics with when forwarding to Datadog, by default, to match the behavior of logs.

Also, is that the complete Helm configuration? I expected to see the Vector configuration in it, but I just see the HAProxy configuration.

**Comment by jonwinton** [comment]
@jszwedko yeah, tried setting it to `false` and it didn't change anything.

That is the full configuration other than the Vector configuration in the original issue body, which I'll re-paste below:

```
data_dir: /vector-data-dir

api:
  enabled: true
  address: 127.0.0.1:8686
  playground: false

sources:
  internal_logs:
    type: internal_logs

  # Datadog Agent telemetry
  datadog_agent:
    type: datadog_agent
    address: "0.0.0.0:6000"
    multiple_outputs: true # To automatically separate metrics and logs
    store_api_key: false

sinks:
  console:
    type: console
    inputs:
      - internal_logs
    target: stdout
    encoding:
      codec: json

  # Datadog metrics output
  datadog_metrics:
    type: datadog_metrics
    inputs:
      - <inputs>...
    api_key: "${DATADOG_API_KEY}"
```

**Comment by neuronull** [comment]
> In `v0.24.0` #13406 contained a fix in the `datadog_metrics` sink.

Just correcting myself- That change I was pointing out was included in v0.23.0, so that whole line of inquiry which followed would be invalid.

**Comment by neuronull** [comment]
Hi @jonwinton,

I've been trying to reproduce this locally and am not having much luck so far. :disappointed: 
Will summarize what I've tried and some observations, and maybe something will jump out at you.

**Additional questions for you:**
1. Do you have any additional log snippets you could provide? For example, is Vector logging any Errors (not debug logs). And, perhaps the error logs from HAProxy?
2. Since we're having trouble reproducing it, would you be open to trying some nightly builds between v0.23.3 and 0.24.0, and observing the first build which exhibits the issue? This could give us hints to which commits might have caused the issue and additionally raise confidence in the issue.
3. Have you attempted this upgrade from 0.23.3 to 0.24.0 multiple times or just once?
4. Have you performed similar upgrades of Vector in this environment before?

**General observations:**
- The message "error shutting down IO" from the Vector logs, seems to be either (likely I suspect) benign or it's not a new issue with 0.24.0, or not specifically tied to the 504 error observed in the Agent logs.
   - This is a debug level print coming from the `hyper` crate
   - I saw this message in every iteration I tried, regardless of whether I upgraded/restarted the Vector pod.
   
- So far I haven't been able to get a 504 error to show in the Agent logs
   - I've seen a 503 error a couple of times, between when Vector pod was uninstalled, and when it was re-installed, but it occurred only once in each case and went away after Vector was running again, and didn't show up every time. 503 while Vector is down seems pretty expected.
   
**My setup:**
- minikube
- Agent using the helm values outlined below.
- Vector using the helm values outlined below.
- A dogstasd client using the program outlined below.
   
**Permutations I've tried:**
- vector 0.23.3 -> vector 0.24.0
- vector 0.23.3 restart vector
- vector 0.24.4 restart vector
- all of the above with and without restarting the Agent pod
- variations on the dogstastd client to adjust the emission rate of distribution metrics and number of distribution metrics.
- variations on the delay between uninstalling vector and re-installing vector

**dogstastd program**
```
from datadog import initialize, statsd
import time
import random
import sys


def log(msg):
    sys.stdout.write('{}: {}\n'.format(time.strftime("%H:%M:%S"), msg))
    sys.stdout.flush()

options = {
    'statsd_host':'172.17.0.3',
    'statsd_port':8125
}

log("initializing")

initialize(**options)

log("initialized")

while (1):
    log("distribution")

    for n in range(0,99):
        name = "foo_metric.distribution_" + str(n)
        value = random.randint(0,999)
        statsd.distribution(name, value, tags=["environment:baz"])

    time.sleep(1)
```

**Agent helm values**
```
datadog:
  apiKey: <redacted>
  containerExclude: "name:vector"
  logs:
    enabled: true
    containerCollectAll: true
  apm:
    enabled: false
clusterAgent:
  enabled: false
agents:
  useConfigMap: true
  image:
    tag: 7.39.1
  customAgentConfig:
    kubelet_tls_verify: false
    vector:
      logs:
        enabled: true
        url: "http://vector-haproxy.default:8282"
      metrics:
        enabled: true
        url: "http://vector-haproxy.default:8282"
    dogstatsd_non_local_traffic: true

ports:
  - containerPort: 8125
    hostPort: 8125
    name: dogstatsdport
    protocol: UDP
```

**Vector helm values**
```
Effectively the same as your provided values , with the exception of tagging specific images, and hard coding the DD API key.
```

**Comment by neuronull** [comment]
> I will definitely do this.

Thanks a bunch!

> One other piece of context that might be helpful: this issue first appeared in our largest staging environment, so I'm wondering if it's related to volume. Are y'all running load testing benchmarks in CI? Screenshot to show the load where we're first encountering the error.

We _some_ sense of load testing within our soak regression tests- generally those soak tests will try to hit vector with the allocated resources available. But we don't have explicit testing to evaluate usage under different loads.

It is good to know that it might be volume related- I can take that into consideration for repro efforts going forward.

**Comment by jonwinton** [comment]
So sorry for the delay @neuronull, but I'm working through the nightlies now. Will get you a report this week so it's ready for the new year 🙇

**Comment by jonwinton** [comment]
@neuronull ok here are the results:

- I found that the regression occurs with the `nightly-2022-08-17-distroless-libc` tag
- [The changes for that build can be found here](https://github.com/vectordotdev/vector/commits/master?after=a7590f98833c2cfc322a6f122155dd8b337f94fe+864&branch=master&qualified_name=refs%2Fheads%2Fmaster)
- I know we've documented it before, but it's the same behavior as before once that nightly tag lands
![Screenshot 2022-12-27 at 2 27 48 PM](https://user-images.githubusercontent.com/4906712/209728007-91e62b12-df80-4325-928b-086fca822b41.png)

**Comment by jszwedko** [comment]
Thanks for bisecting that down @jonwinton ! Just to confirm, do I take the above to mean that `nightly-2022-08-16-distroless-libc` did not exhibit the issue?

**Comment by jonwinton** [comment]
@jszwedko correct! Here's a screenshot showing the `version` was `nightly-2022-08-16-distroless-libc` right before deploying the `2022-08-17` tag and everything was working as expected.

I'm working with DataDog Support to give you access to a notebook with the entire breakdown, but screenshots for now.

<img width="1141" alt="Screenshot 2022-12-27 at 3 06 38 PM" src="https://user-images.githubusercontent.com/4906712/209735123-4513a433-8a3b-404a-87c6-589b4148fb78.png">

**Comment by jszwedko** [comment]
Thanks @jonwinton ! That would seem to narrow down the suspicious commits to https://github.com/vectordotdev/vector/compare/ba1e1c4...01c7a36 then. Nothing immediately jumps out, but it should give us a much better place to start investigating from. The person investigating this is OOO this week, but we'll pick it back up in the new year. Appreciate all of the added context!

**Comment by jonwinton** [comment]
Sounds good! Yeah I was beginning to look at https://github.com/vectordotdev/vector/pull/13973 since we only see the errors come up for distribution data, but I need a Rust pro to help out. That said, if any tagged releases can be pushed out that have any changes to test in our env, I can easily deploy them. 

See y'all next year!

**Comment by neuronull** [comment]
Thanks for going through that effort and providing all that data @jonwinton , really helps!

I believe the hdrhistogram crate updated in #13973 can be ruled out as it only seems to be used in some example code.

After looking at that commit range, the one that stands out most suspiciously is #13960.

At first glance it looks like its unrelated because the code is only operating over `Counter` metric types, and here the affect type is distribution.

But then I noticed this in the helm values ...

```
"collect_counters_with_distributions": true
```

I'm currently working on finding out what that actually translates to in terms of the Agent's behavior.

But so far that commit is really the only one that jumps out to me.

We could also try a test of a private build containing the commits from 8-17 but minus this one.

**Comment by jonwinton** [comment]
@neuronull dang, nice digging! 🙇 

A private image works, or if you can give me the build commands for generating the libc image I can push an image into our private ECR repo. I tried digging into the build pipeline a bit, but my lack of Rust familiarity is holding me back a bit.

**Comment by neuronull** [comment]
Thanks @jonwinton !

Here are the commands to get to the correct state of the repo:

1. git checkout 01c7a36
2. git revert ef22eb9
2.a To resolve the conflict, open lib/vector-core/Cargo.toml and you want the line 89 to be `test = ["vector-common/test"]`

To build the docker image to your local cache:

1. make package-deb-x86_64-unknown-musl
1.a (you can select different architectures too, see the available ones in Makefile)
2. PUSH=false ./scripts/build-docker.sh

Let us know how that goes or if you have any problems! Very curious to hear the result.

**Comment by neuronull** [comment]
@jonwinton , sorry for the spam- Ignore that last comment's instructions.
We're working on improving the procedure to follow.
In the meantime, I'll create the image and push it to the vector repo for you.

**Comment by jonwinton** [comment]
Perfect! I'll test this now

**Comment by jonwinton** [comment]
@neuronull confirmed that we don't see the same issue with this version!

**Comment by jonwinton** [comment]
Oh dang, looking at that PR more, we're interested in maintaining the fix for this bug: https://github.com/vectordotdev/vector/issues/13870

We're also dealing with `interval` issues and this would be helpful once we can safely upgrade 😬

**Comment by neuronull** [comment]
Great to hear that the build had the expected outcome!

Yes, the next steps are for us to figure out what is wrong with that commit :)

We'll definitely want to maintain the fix functionality. Good to know you also require that. Will keep this thread posted on progress~

**Comment by neuronull** [comment]
Hi @jonwinton , I believe this issue may have been solved by  #14741 , which was included in v0.24.2.

Would you mind trying that build?

I think we just didn't realize that it might have been already fixed in a patch release, because on the surface it looked so unrelated to any fixes in v0.24.1 or v0.24.2.

**Comment by jonwinton** [comment]
Hmm I can retry later, but I went as high as 0.25.0 and still had the issue...I will confirm tomorrow

**Comment by neuronull** [comment]
Interesting 🤔 . Ok thank you 🙇

**Comment by jonwinton** [comment]
@neuronull confirmed it's still borked in the `v0.26.0` release 😭

**Comment by neuronull** [comment]
Thanks for the follow-up on that @jonwinton !

In looking at your screenshots, looks like you have vector running in multiple pods.
I'm just curious if you could describe how many instances you have of the Datadog Agent running, and how many instances of Vector running?

I'm investigating another issue that may overlap with this one. Thanks!

**Comment by neuronull** [comment]
Would you also be willing to share your HAProxy config ?

**Comment by jonwinton** [comment]
@neuronull the HAProxy config [is here](https://github.com/vectordotdev/vector/issues/15292#issuecomment-1322465357), but pasting below without the Vector config:

```
haproxy:
  enabled: true
  image:
    tag: &version 2.6.6
  resources:
    requests:
      cpu: 1000m
      memory: 1Gi
    limits:
      cpu: 1000m
      memory: 1Gi

  autoscaling:
    enabled: true
    minReplicas: 9
    maxReplicas: 50

  podLabels:
    tags.datadoghq.com/service: vector-haproxy
    tags.datadoghq.com/version: *version

  affinity:
    podAntiAffinity:
      preferredDuringSchedulingIgnoredDuringExecution:
        - podAffinityTerm:
            labelSelector:
              matchExpressions:
              - key: app.kubernetes.io/name
                operator: In
                values:
                - vector
            topologyKey: failure-domain.beta.kubernetes.io/zone
          weight: 100

  podAnnotations:
    ad.datadoghq.com/haproxy.checks: |
      {
        "haproxy": {
          "init_config": {},
          "instances": [{
            "use_openmetrics": true,
            "openmetrics_endpoint": "http://%%host%%:1024/metrics",
            "histogram_buckets_as_distributions": true,
            "collect_counters_with_distributions": true
          }]
        }
      }

  customConfig: |
    global
      log stdout format raw local0
      maxconn 4096
      stats socket /tmp/haproxy
      hard-stop-after {{ .Values.haproxy.terminationGracePeriodSeconds }}s

    defaults
      log     global
      option  dontlognull
      retries 10
      option  redispatch
      option  allbackups
      timeout client 10s
      timeout server 10s
      timeout connect 5s

    resolvers coredns
      nameserver dns1 kube-dns.kube-system.svc.cluster.local:53
      resolve_retries 3
      timeout resolve 2s
      timeout retry 1s
      accepted_payload_size 8192
      hold valid 10s
      hold obsolete 60s

    frontend stats
      mode http
      bind :::1024
      option httplog
      http-request use-service prometheus-exporter if { path /metrics }
      stats enable
      stats hide-version  # Hide HAProxy version
      stats realm Haproxy\ Statistics  # Title text for popup window
      stats uri /haproxy_stats  # Stats URI
      stats refresh 10s


    frontend datadog-agent
      mode http
      bind :::6000
      option httplog
      option dontlog-normal
      default_backend datadog-agent

    backend datadog-agent
      mode http
      balance leastconn
      option tcp-check
      server-template srv 10 _datadog-agent._tcp.{{ include "vector.fullname" $ }}.{{ $.Release.Namespace }}.svc.cluster.local resolvers coredns check
```

> 'm just curious if you could describe how many instances you have of the Datadog Agent running, and how many instances of Vector running?

So this comes up in our staging environment where we have the following:

- ~150 Nodes all running a DataDog Agent 
- ~20 Vector pods running on v0.23.3. This is usually less but we're scaled out a bit for testing some transforms.
- 9 HAProxy pods

When we deploy any SHA/version beyond v0.23.3, the HPA on Vector tries to run 100+ Vector pods and we still see requests consistently failing even when HPA scales out.

**Comment by jonwinton** [comment]
@neuronull it's hard to tell because anytime we go beyond `v0.23.3` we see such a huge increase in errors that we see reductions across the board. Because it takes a certain amount of traffic from the DD Agent --> Vector to notice this, I have to deploy into a env where our engineers are working and so I can never keep it on for too long.

> Specifically, count and rate ?

So the fun thing about all this: we're run an entirely Prometheus-based ecosystem and so we don't submit any `rate` metrics, only `count`/`gauge`/`historgram`/`summary` types.

---

But regarding your question, looking at the test time from yesterday with `v0.26.0`, these error charts show error counts at the HAProxy layer and show that the Vector backend begins having issues responding. The 2xx rate drops so I'm assuming the requests are retried and make it through OR they're dropped entirely. But since I don't have a great way of identifying which `count` metrics decrease.

<img width="980" alt="Screenshot 2023-01-10 at 8 05 05 AM" src="https://user-images.githubusercontent.com/4906712/211601689-a57eebbe-a575-4d32-a2f1-aaeecd39485f.png">

This is a screenshot of our of our most popular `count` metrics. While we do see an initial drop in datadpoints, it seems to get back to normal as the number of Vector pods scales out.

<img width="1325" alt="Screenshot 2023-01-10 at 7 59 41 AM" src="https://user-images.githubusercontent.com/4906712/211600371-71e3b497-d489-49e3-8bf3-bc810e14b597.png">


Does that help at all? Not a great answer, but I'm hoping some of that was useful

**Comment by neuronull** [comment]
Hi @jonwinton , I have a build that contains a fix candidate. 
Because I've yet to be able to repro this locally, it's a bit of a shot in the dark.
But it essentially is a partial revert of some of the changes that went into that commit we identified earlier, while still maintaining the desired behavior of aggregation of metrics.

This is the image:  `timberio/vector:9c2662cfa_01-13-2023-distroless-libc`

If you wouldn't mind testing that out, it'd be appreciated!

Note that this one is based from `v0.25.2`

**Comment by jonwinton** [comment]
Of course! Let me check this now!

**Comment by jonwinton** [comment]
@neuronull unfortunately the issue is still around. Seems like 


![Screenshot 2023-01-12 at 12 46 17 PM](https://user-images.githubusercontent.com/4906712/212177493-ed9d7128-aaa8-4880-bf2e-225f2626f9a7.png)
![Screenshot 2023-01-12 at 12 48 12 PM](https://user-images.githubusercontent.com/4906712/212178100-01e5068c-dce3-47df-b20b-c9ae5fa3a01e.png)

I definitely did see missing count metrics while this version was deployed but when I rolled back the test the values did fill in. The data seemed to be buffered in the DD Agents and then backfilled once I began the rollback.

**Comment by neuronull** [comment]
Thanks so much for the quick turn around.
And good to know that you also observed a decrease in count metrics.

I just found another potential problem. Unfortunately as I haven't been able to repro locally, your efforts to try builds are currently our best method of narrowing down the root cause.

If you don't mind another, I'm prepping a build now...

**Comment by jonwinton** [comment]
Not at all! Bring them on 👍

**Comment by neuronull** [comment]
`timberio/vector:86635f66b_2023-01-12-distroless-libc`

🤞

**Comment by jonwinton** [comment]
@neuronull sorry for the delay on that! Same results as last time, but I was able to grab a screenshot showing some missing counter metrics for the same charts that we've been referencing:


![Screenshot 2023-01-13 at 9 03 10 AM](https://user-images.githubusercontent.com/4906712/212377128-5bfdfdff-9d32-4f49-b025-3b4e6a3d3497.png)

---

After rollback the data does fill in:

![Screenshot 2023-01-13 at 9 05 52 AM](https://user-images.githubusercontent.com/4906712/212377635-57d9255d-1a91-40e0-8043-04744523cec1.png)

But in general, same results

**Comment by neuronull** [comment]
No worries! Thanks for providing that additional data about the missing count metrics.
This is helpful. I'll regroup here and see about next steps. I'm hopeful since we have been able to identify the problematic commit. Will keep you posted~

**Comment by neuronull** [comment]
To re-confirm @jonwinton, you're using dogstatsd as  an ingestion method, is that correct?

**Comment by jonwinton** [comment]
@neuronull we use in-app Prometheus clients to generate metrics that are then collected with the DataDog Agent OpenMetrics integration ([docs](https://docs.datadoghq.com/integrations/openmetrics/)). Then The DD Agent forwards them (though not sure the exact format) to Vector ([docs](https://docs.datadoghq.com/observability_pipelines/integrations/integrate_vector_with_datadog/?tab=yaml#send-metrics-to-vector)).

**Comment by neuronull** [comment]
Hi @jonwinton,

We are theorizing now about the high volume of your environment and potentially that the aggregation algorithm might not be scaling well with that, possibly causing these issues observed.
Would it be feasible for you to over provision Vector? like at 2x or 2.25x what it was from the recent tests. In the last screen shot if there were roughly 60 from the autoscaling, try 120 or 130.
This would help us understand if that theory has legs and if it does then we can shift focus to optimizing it.

**Comment by neuronull** [comment]
> yeah I can do that, but it might need to wait until tomorrow. 

Thanks! No worries.

> Just to confirm the test: we want to run it at steady state for a set amount of time scaled up to see if the CPU usage and error rates return to normal?

Yes, the key being to over provision by roughly 2x. If it auto scales up beyond that that is ok but the idea is exactly like you said, see if the CPU usage and errors / metric hits return to normal with having 2x or more Vector instances.

> Also, would it be possible to backport the interval fix (https://github.com/vectordotdev/vector/issues/15292#issuecomment-1372477689) while we continue to work on this? The interval issue is causing a lot of issues for us and it would be great to have a fix for that 🤞

What release are you looking to have that backported into ? `v0.23` ?  cc @jszwedko

**Comment by jonwinton** [comment]
> What release are you looking to have that backported into ? `v0.23` ?

Yeah that would be perfect!

**Comment by neuronull** [comment]
We're taking this opportunity to create a process for this kind of request.
Wherein, we would like to publish a build based off of a branch, but don't need to go through the complete patch release cycle.
As it is, it's a manual process and delicate to get a system into the correct configuration to generate the build.
(creating the branch with the necessary changes is trivial).

So we are working on making that a github action, to utilize the CI infra we have.
Then we could essentially automate the build process down to just a few button clicks, once a git branch is pushed to remote.

Currently shooting for next week to accomplish that 🙏

**Comment by neuronull** [comment]
An update @jonwinton  - we will still be pursing the automation of these one off builds, but won't be able to get to it in the short term so I'll go ahead and spin a build that backports the interval fix to the last patch release of v0.23 in the meantime. Hopefully can get that today, if not, early next week.

Are you still planning to try out over provisioning vector? Cheers~

**Comment by neuronull** [comment]
Ok, here is the build. It is from this branch:

https://github.com/vectordotdev/vector/compare/v0.23.3...neuronull/backport_dd_metrics_interval_fix

```
timberio/vector:2023-01-27_backport_dd_metrics_interval_fix-distroless-libc
```

**Comment by jonwinton** [comment]
🙇 Sorry for the delay @neuronull ! I appreciate backporting this 🙇 

I'm definitely still planning to 2x the env and run the test, just got caught up in a migration and went heads down on that. I'll be able to come back to this on Wednesday or Thursday of this week!

**Comment by jonwinton** [comment]
Ok! Working on this now!

**Comment by neuronull** [comment]
Thanks a bunch @jonwinton !
This is essentially what we expected to see.

I'll dive into the performance of that algorithm.

**Comment by neuronull** [comment]
> I'll dive into the performance of that algorithm.

FYI, this is queued for next week, as I have other priorities in flight.

**Comment by neuronull** [comment]
Hi @jonwinton , apologies for the lengthy silence!

I have a fix candidate build available, which contains optimizations to that algorithm.
The performance varies depending on the shape of the input data, but the improvements are up to roughly 66% in some cases.

Very curious to hear if this alleviates the issues you are seeing. Would you mind testing it out when you have time?

`timberio/vector:2023-03-03_optimize_dd_metrics_aggregation-distroless-libc`

**Comment by jonwinton** [comment]
@neuronull amazing! Thanks for this! I'm going oncall for our team tomorrow and will definitely test it out then 😬

**Comment by neuronull** [comment]
> @neuronull amazing! Thanks for this! I'm going oncall for our team tomorrow and will definitely test it out then 😬

Hey @jonwinton ! Have you had a chance to try that out? Very eager to hear the results 🤓 .

**Comment by jonwinton** [comment]
Hey! Sorry for that delay, the DD outage last week threw off a lot of things 😬 

So some results...

I bumped HPA to 200 before the test so that we'd start with a nigh number of pods. Once the new image was deployed, CPU was double that of the old implementation. It took scaling to 400 pods for there to be equity between the two versions in terms of resource utilization. At the tail end, I bumped to 600 pods and we can see it drops a little more.

<img width="1526" alt="Screenshot 2023-03-16 at 11 56 08 AM" src="https://user-images.githubusercontent.com/4906712/225724916-40d4fca4-50e5-4564-98f7-ea44964e10ae.png">

But during this time, error rates stayed elevated in terms of 😭 Datadog Agent --> Vector. I even bumped HAProxy from 9 to 40 pods.

<img width="1013" alt="Screenshot 2023-03-16 at 11 55 29 AM" src="https://user-images.githubusercontent.com/4906712/225724666-6047acb5-ef60-49af-8cf9-30be44dfbcb6.png">

It's important to keep in mind that in this env, we only run 20 vector pods right now

**Comment by neuronull** [comment]
Whoa... intriguing results! The shape of the input data definitely has an effect on the algorithm's performance, so one thing we might want to do is get a sample of data, if that were possible, so that I can profile the function with that input data.

But before that, a couple of things to check- if CPU throttling is going on (that would be reported in a `cpu.throttled` metric in Datadog. Additionally, do you have the env var `VECTOR_THREADS` set, or the `--threads` arg passed in for vector?

**Comment by neuronull** [comment]
Another thing to try- you could set VECTOR_THREADS to match the value in your `resources.limits.cpu`.

**Comment by neuronull** [comment]
Hi @jonwinton !

Aside from the above queries/settings in Vector to try, if those don't improve anything, we can try capturing some data from the Agent log, to see what shape of data is coming into Vector. This should be an easier thing to try than plugging in a packet capture between the Agent and Vector.

The steps are basically:
- add this setting to your Agent config: `log_payloads: true`.
- enable Debug log level in the Agent config: `log_level: debug`.
- restart the Agent.
- If there is a ton of volume ingested by this Agent, then the debug log level could fill the max log size pretty quickly. Depending on your environment, you can either specify a log file in the Agent config `log_file:  ./my_agent_debug.log`
, or you can[ send an Agent Flare to Datadog](https://docs.datadoghq.com/agent/troubleshooting/send_a_flare/?tab=agentv6v7).

This _might_ give me enough to work with, without having to go to the more involved step of a packet capture.

Thanks again for all your assistance and patience here!

**Comment by 4wdonny** [comment]
We're running into what I believe is also this issue. Following along for updates.

**Comment by aashery-square** [comment]
@neuronull 👋 I've picked up this bug from Jon Winton at Cash. Since it looks like there's a fix out for this, would you be able to provide us with a test image containing the fix that we can demo?
